compiler/lib/linker.ml

```diff
@@ -396,20 +416,45 @@ let load_fragment ~filename f =
                 let named_values = find_named_value code in
                 Primitive.register name kind ka arity;
                 StringSet.iter Primitive.register_named_value named_values;
-                (if Hashtbl.mem provided name
-                then
-                  let _, ploc, weakdef = Hashtbl.find provided name in
-                  if not weakdef
+                let is_updating =
+                  if Hashtbl.mem provided name
                   then
-                    warn
-                      "warning: overriding primitive %S\n  old: %s\n  new: %s@."
-                      name
-                      (loc ploc)
-                      (loc pi));
-                Hashtbl.add provided name (id, pi, weakdef);
-                Hashtbl.add provided_rev id (name, pi);
-                check_primitive ~name pi ~code ~requires;
-                Hashtbl.add code_pieces id (code, requires)))
+                    let _, ploc, weakdef, prev_target_env = Hashtbl.find provided name in
+                    if not weakdef
+                    then (
+                      match prev_target_env, target_env with
+                      | `Isomorphic, `Browser -> Config.Flag.is_targetting_browser_env ()
+                      | `Isomorphic, `Nodejs -> Config.Flag.is_targetting_nodejs_env ()
+                      | `Nodejs, `Browser
+                      | `Browser, `Nodejs
+                      | `Nodejs, `Isomorphic
+                      | `Browser, `Isomorphic ->
+                          warn
+                            "warning: target_env should not transition from one \
+                             specialization to another. ignoring. %S\n\
+                            \  old: %s\n\
+                            \  new: %s@."
+                            name
+                            (loc ploc)
+                            (loc pi);
+                          false
+                      | _ ->
+                          warn
+                            "warning: overriding primitive %S\n  old: %s\n  new: %s@."
+                            name
+                            (loc ploc)
+                            (loc pi);
+                          true)
+                    else true
```

Comment:
Putting the smaller branch of an if-then-else at the top usually helps review. `If b then many_lines_long else true` -> `if not b then true else many_lines_long`

Reply:
ya, you're right. i didn't realize this. thank you for the tip.

14c237160a9d9eb4cfad87571167784fce893973